### ISSUE TITLE
Bad Perf on Softmax Fusion for Batch Size 1

### ISSUE THREAD

**Issue by kevinstephano** [csarofeen/pytorch#880]
## 🐛 Bug
The Softmax fusion in Vision Transformer is performing much worse than the eager mode equivalent.  This is probably a heuristic issue.

```
Start(sec)  Duration(nsec)  CorrId  GrdX  GrdY  GrdZ  BlkX  BlkY  BlkZ  Reg/Trd  StcSMem  DymSMem  Bytes  Thru(MB/s)  SrcMemKd  DstMemKd        Device        Ctx  Strm                                                  Name                                                
 ----------  --------------  ------  ----  ----  ----  ----  ----  ----  -------  -------  -------  -----  ----------  --------  --------  ------------------  ---  ----  ----------------------------------------------------------------------------------------------------
   2.107839           11648     902   864     1     1   256     1     1       27        0        0                                         A100-SXM4-40GB (0)    1     7  _ZN2at6native89_GLOBAL__N__65_tmpxft_00001610_00000000_10_DistributionNormal_compute_86_cpp1_ii_7d8…
   2.133106            6240     908  1213     1     1    64     1     1       16        0        0                                         A100-SXM4-40GB (0)    1     7  void at::native::vectorized_elementwise_kernel<4, at::native::MulScalarFunctor<float, float>, at::d…
   2.133217            7615     917   394     1     1    32     4     1       29        0        0                                         A100-SXM4-40GB (0)    1     7  void (anonymous namespace)::softmax_warp_forward<float, float, float, 8, false>(float*, float const…
   2.457719           28927     948   226     1     1    25     7     1       26        0      700                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, 4>, double, CudaCodeGen::Tensor<float, 4>)          
   2.457912           26207     955   226     1     1    25     7     1       26        0      700                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, 4>, double, CudaCodeGen::Tensor<float, 4>)          
   2.458597           26336     962   226     1     1    25     7     1       26        0      700                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, 4>, double, CudaCodeGen::Tensor<float, 4>)          
   2.458628           26304     969   226     1     1    25     7     1       26        0      700                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, 4>, double, CudaCodeGen::Tensor<float, 4>) 
```

## To Reproduce
```
import torch
from torch import nn
from torch.nn import Module
import torch.nn.functional as F

import math

torch._C._jit_set_nvfuser_enabled(True)
torch._C._jit_set_texpr_fuser_enabled(False)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
torch._C._jit_override_can_fuse_on_cpu(False)
torch._C._jit_override_can_fuse_on_gpu(False)
torch._C._jit_set_bailout_depth(20)

class BertConfig :
    def __init__(self) :
        self.hidden_size = 1024
        self.num_attention_heads = 16
        self.hidden_dropout_prob = 0.1
        self.attention_probs_dropout_prob = 0.1
        self.num_layers = 10

class Fusion(nn.Module):
    def __init__(self, config):
        super(Fusion, self).__init__()
        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)

    def forward(self, inputs):
        out1 = inputs / math.sqrt(self.attention_head_size)
        out3 = F.softmax(out1, dim=-1)
        out4 = self.dropout(out3)
        return out4

if __name__ == "__main__" :
    inputs = torch.randn(1, 8, 197, 197, device="cuda", dtype=torch.float, requires_grad=False)
    #mask = torch.randn(1, 1, 1, 197, device="cuda", dtype=torch.float, requires_grad=False)
    #mask_bool = mask > 0.
   
    model = Fusion(BertConfig())
    model.cuda()
    model.eval()
   
    jit_model = torch.jit.script(model)
  
    for idx in range(5) :
        out = jit_model(inputs)
```

Something that is interesting is that the Vision Transformer model doesn't have a mask but Bert does.  When I had the mask back in, the fusion's perf is better as it changes the blocking.  The block 150 threads instead.

```
 Start(sec)  Duration(nsec)  CorrId  GrdX  GrdY  GrdZ  BlkX  BlkY  BlkZ  Reg/Trd  StcSMem  DymSMem  Bytes  Thru(MB/s)  SrcMemKd  DstMemKd        Device        Ctx  Strm                                                  Name                                                
 ----------  --------------  ------  ----  ----  ----  ----  ----  ----  -------  -------  -------  -----  ----------  --------  --------  ------------------  ---  ----  ----------------------------------------------------------------------------------------------------
   2.018881           11616     902   864     1     1   256     1     1       27        0        0                                         A100-SXM4-40GB (0)    1     7  _ZN2at6native89_GLOBAL__N__65_tmpxft_00001610_00000000_10_DistributionNormal_compute_86_cpp1_ii_7d8…
   2.019484            5504     921     1     1     1   256     1     1       27        0        0                                         A100-SXM4-40GB (0)    1     7  _ZN2at6native89_GLOBAL__N__65_tmpxft_00001610_00000000_10_DistributionNormal_compute_86_cpp1_ii_7d8…
   2.019701            4736     934     1     1     1    64     1     1       16        0        0                                         A100-SXM4-40GB (0)    1     7  void at::native::vectorized_elementwise_kernel<4, at::native::BUnaryFunctor<at::native::CompareGTFu…
   2.043698            6080     941  1213     1     1    64     1     1       16        0        0                                         A100-SXM4-40GB (0)    1     7  void at::native::vectorized_elementwise_kernel<4, at::native::MulScalarFunctor<float, float>, at::d…
   2.046094           12192     947  1213     1     1    64     1     1       32        0        0                                         A100-SXM4-40GB (0)    1     7  void at::native::unrolled_elementwise_kernel<at::native::AddFunctor<float>, at::detail::Array<char*…
   2.046197            7585     956   394     1     1    32     4     1       29        0        0                                         A100-SXM4-40GB (0)    1     7  void (anonymous namespace)::softmax_warp_forward<float, float, float, 8, false>(float*, float const…
   2.386294           18176     987   526     1     1    50     3     1       27        0      600                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<bool, 4>, CudaCodeGen::Tensor<float, 4>, double, CudaCodeG…
   2.386472           16576     994   526     1     1    50     3     1       27        0      600                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<bool, 4>, CudaCodeGen::Tensor<float, 4>, double, CudaCodeG…
   2.387091           16384    1001   526     1     1    50     3     1       27        0      600                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<bool, 4>, CudaCodeGen::Tensor<float, 4>, double, CudaCodeG…
   2.387122           16512    1008   526     1     1    50     3     1       27        0      600                                         A100-SXM4-40GB (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<bool, 4>, CudaCodeGen::Tensor<float, 4>, double, CudaCodeG…
```

```
import torch
from torch import nn
from torch.nn import Module
import torch.nn.functional as F

import math

torch._C._jit_set_nvfuser_enabled(True)
torch._C._jit_set_texpr_fuser_enabled(False)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
torch._C._jit_override_can_fuse_on_cpu(False)
torch._C._jit_override_can_fuse_on_gpu(False)
torch._C._jit_set_bailout_depth(20)

class BertConfig :
    def __init__(self) :
        self.hidden_size = 1024
        self.num_attention_heads = 16
        self.hidden_dropout_prob = 0.1
        self.attention_probs_dropout_prob = 0.1
        self.num_layers = 10

class Fusion(nn.Module):
    def __init__(self, config):
        super(Fusion, self).__init__()
        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)

    def forward(self, inputs, mask):
        out1 = inputs / math.sqrt(self.attention_head_size)
        out2 = out1 + mask
        out3 = F.softmax(out2, dim=-1)
        out4 = self.dropout(out3)
        return out4

if __name__ == "__main__" :
    inputs = torch.randn(1, 8, 197, 197, device="cuda", dtype=torch.float, requires_grad=False)
    mask = torch.randn(1, 1, 1, 197, device="cuda", dtype=torch.float, requires_grad=False)
    mask_bool = mask > 0.
   
    model = Fusion(BertConfig())
    model.cuda()
    model.eval()
   
    jit_model = torch.jit.script(model)
  
    for idx in range(5) :
        out = jit_model(inputs, mask_bool)
```
